### PR TITLE
pyupgrade --py37-plus

### PR DESCRIPTION
**Pull request by mmerickel**

### CODE REVIEW

**Review by stevepiercy — APPROVED** [review]
One minor change request, and a few questions out of curiosity.

File: src/pyramid/authentication.py [hunk 1]
```diff
@@ -74,7 +74,7 @@ def authenticated_userid(self, request):
 
         if self.callback is None:
             debug and self._log(
-                'there was no groupfinder callback; returning %r' % (userid,),
+                f'there was no groupfinder callback; returning {userid!r}',
```
[the remark this answers is not in the document]
Comment (bertjwregeer) [reply src/pyramid/authentication.py:77]: That is deliberate. Please read: https://docs.python.org/3/library/string.html#formatspec for more information on the formatting spec/options available.

File: src/pyramid/authentication.py [hunk 2]
```diff
@@ -717,7 +717,7 @@ def digest(self):
         )
 
     def cookie_value(self):
-        v = '%s%08x%s!' % (self.digest(), int(self.time), quote(self.userid))
+        v = f'{self.digest()}{int(self.time):08x}{quote(self.userid)}!'
```
[the remark this answers is not in the document]
Comment (bertjwregeer) [reply src/pyramid/authentication.py:720]: This is padding the bites as necessary, please see https://docs.python.org/3/library/string.html#formatspec for more information on formatting. It should not be outside the closing bracket, it is correct where it is.

File: src/pyramid/scripts/proutes.py [hunk 3]
```diff
@@ -42,7 +42,7 @@ def _get_print_format(fmt, max_name, max_pattern, max_view, max_method):
 
     for index, col in enumerate(fmt):
         size = max_map[col] + PAD
-        print_fmt += '{{%s: <{%s}}} ' % (col, index)
+        print_fmt += f'{{{{{col}: <{{{index}}}}}}} '
```
[the remark this answers is not in the document]
Comment (bertjwregeer) [reply src/pyramid/scripts/proutes.py:45]: Yes, all those brackets are correct. In format strings `{{` means literal `{` in the output. See https://docs.python.org/3/library/string.html#formatspec

**Review by bertjwregeer — APPROVED** [review]

File: src/pyramid/authentication.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (stevepiercy) [reply src/pyramid/authentication.py:77]: Thank you! I found the grammar spec here, which includes the `!`: https://docs.python.org/3/library/string.html#format-string-syntax

File: src/pyramid/encode.py [hunk 4]
```diff
@@ -70,13 +70,13 @@ def urlencode(query, doseq=True, quote_via=quote_plus):
         if is_nonstr_iter(v):
             for x in v:
                 x = quote_via(x)
-                result += '%s%s=%s' % (prefix, k, x)
```
Comment (huntcsg) [inline src/pyramid/encode.py:73]: I love everything about this PR.

But I think the changes in this function specifically end up with less readable code.

I think the sprintf (or maybe the format string style) would be more readable than the f-string style.
```
%s%s=%s
%s%s=
%s%s=%s

{}{}={}
{}{}=
{}{}={}
```

In my opinion these styles cases are easier to read at a glance than:

```
{prefix}{k}={x}
{prefix}{k}=
{prefix}{k}={v}
```

Like I said, I love this PR, but I think in this particular case it ends up with less readable code, not more readable code.

Just 2️⃣ 🪙

Comment (bertjwregeer) [reply src/pyramid/encode.py:73]: I much prefer the f-string equivalent as much easier to read. It doesn't require my eyes jumping back and forth between where the variable is, and what the type is that we are formatting into the string.

Comment (huntcsg) [reply src/pyramid/encode.py:73]: My comment was 100% bike shedding and I love this PR.

Comment (ztane) [reply src/pyramid/encode.py:73]: f-strings and percent formatting are marginally faster than the `''.format()` too. f-strings ftw.

**Pull request merged**